streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -568,6 +568,9 @@ void suspend(final boolean clean,
                 commit(false);
             } finally {
                 if (eosEnabled) {
+                    // In #suspend, we always checkpoint offsets after commit even with EOS turned on
+                    // to avoid race condition when writing checkpoint file through StateManager #closeSuspended.
+                    stateMgr.checkpoint(activeTaskCheckpointableOffsets());
```

Comment:
I seems, we should remove writing the checkpoint file in `closeStateManager`, too? Note, that `suspend()` will be called anyway (during rebalance, and during shutdown)

Also, why do we write the checkpoint if EOS is enabled only? To me, it seems we can write the checkpoint file if EOS is disabled, too?

\cc @guozhangwang

Reply:
Because in the #commit() function we shall write checkpoint file if EOS is not turned on. So we are avoiding writing checkpoint file twice in consecutive.

Reply:
Ack. Might be worth to add this to the comment.

Reply:
Ack. Might be worth to add this to the comment.

Reply:
Sure, will do!

Reply:
What about:

> I seems, we should remove writing the checkpoint file in closeStateManager, too? Note, that suspend() will be called anyway (during rebalance, and during shutdown)

Reply:
I see, will address this!